base/common/src/org/dogtagpki/nss/NSSDatabase.java

```diff
@@ -0,0 +1,219 @@
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License along
+// with this program; if not, write to the Free Software Foundation, Inc.,
+// 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+//
+// (C) 2019 Red Hat, Inc.
+// All rights reserved.
+// --- END COPYRIGHT BLOCK ---
+
+package org.dogtagpki.nss;
+
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.FileWriter;
+import java.io.InputStreamReader;
+import java.io.OutputStreamWriter;
+import java.io.PrintWriter;
+import java.io.Reader;
+import java.io.Writer;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+
+import org.apache.commons.io.FileUtils;
+
+/**
+ * @author Endi S. Dewata
+ */
+public class NSSDatabase {
+
+    public static org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger(NSSDatabase.class);
+
+    File directory;
+
+    public NSSDatabase(File directory) {
+        this.directory = directory;
+    }
+
+    public File getDirectory() {
+        return directory;
+    }
+
+    public void setDirectory(File directory) {
+        this.directory = directory;
+    }
+
+    public boolean exists() {
+        return directory.exists();
+    }
+
+    public void create() throws Exception {
+        create(null);
+    }
+
+    public void create(String password) throws Exception {
+
+        logger.info("Creating NSS database in " + directory);
+
+        directory.mkdirs();
+
+        File passwordFile = new File(directory, "password.txt");
+
+        try {
+            List<String> command = new ArrayList<>();
+            command.add("certutil");
+            command.add("-N");
+            command.add("-d");
+            command.add(directory.getAbsolutePath());
+
+            if (password == null) {
+                command.add("--empty-password");
+
+            } else {
+                try (PrintWriter out = new PrintWriter(new FileWriter(passwordFile))) {
+                    out.println(password);
+                }
+
+                command.add("-f");
+                command.add(passwordFile.getAbsolutePath());
+            }
+
+            debug(command);
+
+            ProcessBuilder pb = new ProcessBuilder(command);
+            pb.inheritIO();
+
+            Process p = pb.start();
+            int rc = p.waitFor();
+
+            if (rc != 0) {
+                throw new Exception("Command failed: rc=" + rc);
+            }
+
+        } finally {
+            if (passwordFile.exists()) passwordFile.delete();
+        }
+
+        // Install p11-kit-trust module if it doesn't exist
+        if (!isModuleInstalled("p11-kit-trust")) {
+            addModule("p11-kit-trust", "/usr/share/pki/lib/p11-kit-trust.so");
```

Comment:
The last remaining thing I want a discussion on is this: automatically adding root system store to the client.

There's an argument that, by only explicitly adding certificates of CAs you trust (namely, only the Dogtag CA -- that's generally the only CA necessary for PKI interactions), you're safer here. It probably only matters on say, KRA operations, and matters less on say, CSR submissions.

I'm not inclined to care too much here; we can assume this is a fairly safe list of CA certs. However, I _am_ inclined to ask @ladycfu if there's any wording that CC perhaps might care about this that we'll need to explicitly mention. Thoughts?

(If they do care, I'd prefer to make this a separate command, something like `pki trust-root-store` perhaps).

Reply:
As discussed on IRC, the p11-kit-trust will only be added to a new NSS database created by the CLI. If the user already has an NSS database without the module (e.g. for CC), the CLI will not add the module. Also if the system administrator wants to limit the trusted certs, that can be done centrally on the system, and the CLI should continue to use the p11-kit-trust to trust the system-provided certs.